lib/ex_unit/lib/ex_unit/case.ex

```diff
@@ -223,7 +223,26 @@ defmodule ExUnit.Case do
     end
 
     quote do
-      async = !!unquote(opts)[:async]
+      @ex_unit_default_opts [async: true]
+
+      old_opts = Module.get_attribute(__MODULE__, :ex_unit_opts)
+      new_opts = Keyword.merge(@ex_unit_default_opts, unquote(opts) || [])
+
+      if old_opts && old_opts != new_opts do
```

Comment:
I think we should warn regardless of different options, and just warn if `use ExUnit.Case` was called before.

Reply:
I was afraid that there was situations that it makes sense to call it twice because of this condition: https://github.com/elixir-lang/elixir/pull/9360/files/4aaea84c6dc9ba6db0bfecbc61888efea7c1d2f5#diff-da9b91caf1cd02e59ee50c01929a2058R247

I will change to the way you suggested and check the whole elixir codebase + some personal phoenix projects to make sure that's not the case. Thanks.

Reply:
Changed it and it warned for this test: https://github.com/elixir-lang/elixir/blob/master/lib/ex_unit/test/ex_unit/case_template_test.exs#L40

Investigating a bit its origin I also discovered the origin of the condition I mentioned on the previous comment: https://github.com/elixir-lang/elixir/commit/acf912c2c87758434d76f57276345f3daa699a78

Looks like the idea was to allow composition of `ExUnit.CaseTemplate`s, should we keep my change the way it is then? So it only warns when the use tries to change the options? Or maybe we should not warn at all and come up with another fix for the issue?

Reply:
Yes, perhaps the best solution is to simply allow async to work through multiple calls. Something like this:

    given_async = unquote(opts)[:async]

and then in the less you do:

    unless ... do
      @ex_unit_async false
    end

And after the unless you do:

    if unquote(opts)[:async] do
      @ex_unit_async true
    end

Reply:
Yeah, I thought about doing that, but didn't do because I thought it was that way for a reason, since it is that way from the day it was introduced: https://github.com/elixir-lang/elixir/commit/acf912c2c87758434d76f57276345f3daa699a78#diff-da9b91caf1cd02e59ee50c01929a2058R35-R45

I will investigate a bit, if I don't find anything, I will change to your suggestion and check it out, maybe some tests would fail if it really breaks stuff 🤷‍♂

Reply:
About your suggestion: should we set async to false inside the `unless`, the following would set it to `false`:

```elixir
defmodule Test do
  use ExUnit.Case, async: true
  use ExUnit.Case
end
```

Maybe the behavior most people would expect here would be not to reset `async` to `false` if the option is not given, IDK.

Reply:
Oh, right, nevermind what I said, the unless would filter out. 🤦‍♂

Reply:
Just changed the condition to `if unquote(opts)[:async] do` that way this would reset it to false:

```elixir
defmodule Test do
  use ExUnit.Case, async: true
  use ExUnit.Case, async: false
end
```